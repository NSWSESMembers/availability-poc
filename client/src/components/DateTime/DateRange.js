import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import { DateSelect } from './';

import styles from './styles';

const DateRange = ({ timeSegments, startTime, endTime, selectedDays, onSelect }) => {
  const startDay = moment.unix(startTime).startOf('day');
  const endDay = moment.unix(endTime).endOf('day');

  const startOfWeek = startDay.startOf('isoWeek');
  const endOfWeek = endDay.endOf('isoWeek');
  const difference = endOfWeek.diff(startOfWeek, 'days') + 1;

  const table = [];
  let rows = [];
  for (let i = 0; i < difference; i += 1) {
    const day = startOfWeek.add(i === 0 ? 0 : 1, 'days').unix();
    const dayUnix = 24 * 60 * 60;
    const endOfDay = parseInt(day, 0) + dayUnix;

    let availType = 'NotSpecified';
    if (day < startTime || day > endTime) {
      availType = 'OutOfRange';
    } else {
      const results = timeSegments.filter(
        timeSegment => day <= timeSegment.startTime && endOfDay >= timeSegment.endTime,
      );
      if (results.length > 0) {
        availType = results[0].status;
      }
    }

    rows.push(
      <DateSelect
        key={`${i}column`}
        availType={availType}
        day={day}
        onSelect={onSelect}
        select={selectedDays.indexOf(day) !== -1}
      />,
    );

    if ((i + 1) % 7 === 0) {
      table.push(
        <View key={`${i}row`} style={styles.week}>
          {rows}
        </View>,
      );
      rows = [];
    }
  }
  return <View style={{ flexDirection: 'column' }}>{table}</View>;
};

DateRange.propTypes = {
  timeSegments: PropTypes.arrayOf(
    PropTypes.shape({
      startTime: PropTypes.number.isRequired,
      endTime: PropTypes.number.isRequired,
      id: PropTypes.number.isRequired,
      status: PropTypes.string.isRequired,
    }),
  ),
  selectedDays: PropTypes.arrayOf(PropTypes.number),
  startTime: PropTypes.number.isRequired,
  endTime: PropTypes.number.isRequired,
  onSelect: PropTypes.func.isRequired,
};

export default DateRange;
