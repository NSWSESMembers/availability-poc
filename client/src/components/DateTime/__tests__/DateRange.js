import React from 'react';
import renderer from 'react-test-renderer';

import DateRange from '../DateRange';

const noop = () => {};

const testData = {
  startTime: 1,
  endTime: 2,
  id: 1,
  status: 'NOT_SPECIFIED',
};

test('renders correctly', () => {
  const tree = renderer.create(
    <DateRange
      startTime={1}
      endTime={3}
      onSelect={noop}
      selectedDays={[2]}
      timeSegments={[testData]}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
