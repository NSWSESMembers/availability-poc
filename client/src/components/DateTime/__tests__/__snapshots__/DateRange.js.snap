// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        Mo
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        29
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        Tu
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        30
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        We
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        31
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        Th
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        01
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        Fr
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        02
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        Sa
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        03
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "borderColor": "#AAA",
            "borderWidth": 1,
            "flex": 1,
            "margin": 2,
            "padding": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        Su
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#AAA",
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        04
      </Text>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
